File: frontend/src/lib/outboundOrder.ts
// 出库单前端类型 + API。对齐后端 schemas/outbound_order.py。
// 🔴 出库单据/行**无任何售价/成本字段**(契约 §3 红线):出库单是纯仓单,全链路不带售价与成本。
// 出库行本身不存快照,展示字段(name/spec/unit)由后端 join SO 行回填(契约 §1.3 单一源头)。
import { api } from "./api";
import type { Page } from "./catalog";

export type OutboundOrderStatus = "DRAFT" | "ISSUED" | "CANCELLED";

/** 出库单行(读)。快照字段来自后端 join CONFIRMED SO 行(冻结),出库行不复制快照。 */
export interface OutboundOrderLineOut {
  id: number;
  outbound_order_id: number;
  sales_order_line_id: number;
  sku_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  qty: number | string;
}

export interface OutboundOrderOut {
  id: number;
  no: string;
  sales_order_id: number;
  shipment_id: number;
  status: OutboundOrderStatus;
  issued_at: string | null;
  note: string | null;
  updated_at: string;
  // 详情响应附带的溯源展示字段。
  sales_order_no: string;
  shipment_no: string | null;
  container_no: string | null;
}

export interface OutboundOrderListItem {
  id: number;
  no: string;
  sales_order_id: number;
  sales_order_no: string;
  shipment_id: number;
  shipment_no: string | null;
  container_no: string | null;
  status: OutboundOrderStatus;
  line_count: number;
  total_qty: number | string;
  issued_at: string | null;
  created_at: string;
}

/** 建单器数据源(GET /sales-orders/{id}/outboundable-lines)。available_qty = 可发。无售价字段。 */
export interface OutboundableLine {
  sales_order_line_id: number;
  sku_id: number;
  name_snapshot: string;
  spec_text_snapshot: string;
  unit_snapshot: string;
  language: string;
  ordered_qty: number | string;
  inbound_qty: number | string;
  outbound_qty: number | string;
  available_qty: number | string;
}

export interface OutboundOrderDetail {
  order: OutboundOrderOut;
  lines: OutboundOrderLineOut[];
}

/** SO 详情「关联出库单」项(仅 outbound:read 者的 SO 详情响应附带;含 CANCELLED 可追溯;无金额)。 */
export interface RelatedOutboundOrder {
  id: number;
  no: string;
  status: OutboundOrderStatus;
  shipment_id: number;
  shipment_no: string;
  container_no: string | null;
  issued_at: string | null;
}

/** 写入行(引用 SO 行 + 本次出库数量)。 */
export interface OutboundOrderLineIn {
  sales_order_line_id: number;
  qty: number;
  sort_order?: number;
}

/** 建单(柜详情内发起):锚定单一柜 + 单一 SO,一柜内每来源 SO 各一张(后端偏唯一)。 */
export interface OutboundOrderCreateBody {
  shipment_id: number;
  sales_order_id: number;
  note?: string | null;
  lines: OutboundOrderLineIn[];
}

/** 草稿整单保存(携乐观锁基线,冲突 409 → 后端 message)。柜/SO 锚定不可改,仅改行 + 备注。 */
export interface OutboundOrderUpdateBody {
  note?: string | null;
  lines: OutboundOrderLineIn[];
  /** 乐观锁基线 = 打开编辑时的 order.updated_at(对齐 PO / 入库)。 */
  expected_updated_at: string;
}

function qs(p: Record<string, unknown>): string {
  const q = new URLSearchParams();
  Object.entries(p).forEach(([k, v]) => {
    if (v !== undefined && v !== null && v !== "") q.set(k, String(v));
  });
  const s = q.toString();
  return s ? `?${s}` : "";
}

export interface OutboundOrderListFilters {
  status?: string;
  shipment_id?: number;
  sales_order_id?: number;
  keyword?: string;
  page?: number;
  size?: number;
}

export const outboundOrderApi = {
  list: (p: OutboundOrderListFilters) =>
    api.get<Page<OutboundOrderListItem>>(`/api/v1/outbound-orders${qs(p as Record<string, unknown>)}`),
  get: (id: number) => api.get<OutboundOrderDetail>(`/api/v1/outbound-orders/${id}`),
  /** 建单器数据源:某 SO 的可发行 + 每行可发数量(须 outbound:manage)。 */
  outboundableLines: (salesOrderId: number) =>
    api.get<{ items: OutboundableLine[] }>(
      `/api/v1/sales-orders/${salesOrderId}/outboundable-lines`,
    ),
  create: (b: OutboundOrderCreateBody) =>
    api.post<OutboundOrderDetail>("/api/v1/outbound-orders", b),
  update: (id: number, b: OutboundOrderUpdateBody) =>
    api.put<OutboundOrderDetail>(`/api/v1/outbound-orders/${id}`, b),
  confirm: (id: number) => api.post<OutboundOrderDetail>(`/api/v1/outbound-orders/${id}/confirm`),
  revert: (id: number, void_reason?: string | null) =>
    api.post<OutboundOrderDetail>(`/api/v1/outbound-orders/${id}/revert`, {
      void_reason: void_reason ?? null,
    }),
  cancel: (id: number) => api.post<OutboundOrderDetail>(`/api/v1/outbound-orders/${id}/cancel`),
};

File: backend/app/schemas/outbound_order.py
"""出库单 schemas。

🔴红线由设计消除:出库单/行**无任何价格/成本/售价字段**(纯仓单,契约 §1.2/§3),
故读投影无需脱敏工厂。展示快照(name/spec/unit)经 join SO 行取(SO 行冻结单一源头,
出库行不复制)。售价侧在应收(receivables,整表 receivable:read 门控),不经出库 API 回显。
"""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

# ---------- 写入(建单 / 整单保存)----------


class OutboundOrderLineIn(BaseModel):
    """出库行:引用 SO 行 + 本次出库数量。sku 由 service 从 SO 行派生(不采信客户端)。无金额字段。"""
    sales_order_line_id: int
    qty: float = Field(gt=0)


class OutboundOrderCreateIn(BaseModel):
    """基于 CONFIRMED SO + OPEN 柜建一张 DRAFT 出库单。lines 至少一行(空 → 422)。"""
    sales_order_id: int
    shipment_id: int
    note: str | None = None
    lines: list[OutboundOrderLineIn] = Field(min_length=1)


class OutboundOrderUpdateIn(BaseModel):
    """整单保存(PUT):仅 DRAFT;必带乐观锁基线(对齐 PO/入库)。SO/柜 归属不可改(单据身份)。行整表重写。"""
    note: str | None = None
    lines: list[OutboundOrderLineIn] = Field(min_length=1)
    expected_updated_at: datetime


class OutboundOrderRevertIn(BaseModel):
    """撤销出库:作废应收留痕原因(可空)。"""
    void_reason: str | None = None


# ---------- 读投影(无金额字段)----------


class OutboundOrderLineOut(BaseModel):
    id: int
    outbound_order_id: int
    sales_order_line_id: int
    sku_id: int
    qty: float
    name_snapshot: str
    spec_text_snapshot: str
    unit_snapshot: str
    language: str


class OutboundOrderOut(BaseModel):
    id: int
    no: str
    sales_order_id: int
    shipment_id: int
    status: str
    issued_at: datetime | None
    note: str | None
    updated_at: datetime
    # 详情投影附带的柜状态(resolve_order_parties 回填);None = 柜缺失(异常)。
    # 供前端「撤销出库」按钮门禁——柜非 OPEN 时禁用(镜像 41910)。default 兜 model_validate(ORM 无此列)。
    shipment_status: str | None = None

    @classmethod
    def build(cls, order, extra: dict | None = None) -> dict:
        d = cls.model_validate(order, from_attributes=True).model_dump()
        if extra:
            d.update(extra)
        return d


class OutboundOrderListItem(BaseModel):
    id: int
    no: str
    sales_order_id: int
    sales_order_no: str
    shipment_id: int
    shipment_no: str
    container_no: str | None
    status: str
    issued_at: datetime | None
    line_count: int
    total_qty: float
    created_at: datetime


class OutboundableLineOut(BaseModel):
    """SO 可发行(建单器数据源)。无成本/售价字段。available = 可发上限。"""
    sales_order_line_id: int
    sku_id: int
    name_snapshot: str
    spec_text_snapshot: str
    unit_snapshot: str
    language: str
    ordered_qty: float
    inbound_qty: float
    outbound_qty: float
    available_qty: float


class RelatedOutboundOrderItem(BaseModel):
    """SO 详情「关联出库单区」项(含 CANCELLED 并标状态,可追溯)。无金额字段。"""
    id: int
    no: str
    status: str
    shipment_id: int
    shipment_no: str
    container_no: str | None
    issued_at: datetime | None
